### ISSUE TITLE
domish gets confused by spaces in xmlns names

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4503]
|<img alt="michich's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| michich reported|
|-|-|
|Trac ID|trac#4503|
|Type|defect|
|Created|2010-06-14 15:17:56Z|
|Branch|https://github.com/twisted/twisted/tree/domish-xmlns-spaces-4503|

I maintain the package Jabbim in Fedora. It's a Twisted-based XMPP client. Leandro Vázquez Cervantes reported to me a crash in Jabbim which turns out to be a bug in Twisted:
```
	  File "/usr/lib/python2.6/site-packages/twisted/internet/tcp.py", line 140, in doRead
	    return Connection.doRead(self)
	  File "/usr/lib/python2.6/site-packages/twisted/internet/tcp.py", line 463, in doRead
	    return self.protocol.dataReceived(data)
	  File "/usr/lib/python2.6/site-packages/twisted/words/xish/xmlstream.py", line 75, in dataReceived
	    self.stream.parse(data)
	  File "/usr/lib/python2.6/site-packages/twisted/words/xish/domish.py", line 756, in parse
	    self.parser.Parse(buffer)
	  File "/usr/lib/python2.6/site-packages/twisted/words/xish/domish.py", line 774, in _onStartElement
	    e = Element(qname, self.defaultNsStack[-1], attrs, self.localPrefixes)
	  File "/usr/lib/python2.6/site-packages/twisted/words/xish/domish.py", line 404, in __init__
	    self.uri, self.name = qname
	exceptions.ValueError: too many values to unpack
```
After some googling I found the same bug was affecting pyaimt in Debian: http://bugs.debian.org/cgi-bin/bugreport.cgi?bug=583263

It happens when a received stanza contains spaces in a xmlns, e.g.:
```
<data xmlns="http://www.xmpp.org/extensions/xep-0084.html#ns-data ">
```
(notice the space before the closing quotation mark)

I don't know if any standard forbids spaces in xmlns IRIs, but a ValueError is certainly not an acceptable result.

The attached patch fixes it.

Attachments:

* [twisted-words-domish-handle-xmlns-names-with-spaces.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/424/424c39141c58d9887c3c441b7b771bc1d29068c4/2476526051c2c8582922c537da9bf030d6c10bf5.patch) (972 bytes) - added by michich on 2010-06-14 15:18:43Z - domish: handle spaces in xmlns IRIs
* [expattest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/424/424c39141c58d9887c3c441b7b771bc1d29068c4/4536b47c23fca49cb873191b471256f1882102ad.py) (751 bytes) - added by michich on 2010-06-14 15:19:59Z - a short demo of the problem

<details><summary>Searchable metadata</summary>

```
trac-id__4503 4503
type__defect defect
reporter__michich michich
priority__normal normal
milestone__ 
branch__branches_domish_xmlns_spaces_4503 branches/domish-xmlns-spaces-4503
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__words words
keywords__ 
time__1276528676000000 1276528676000000
changetime__1279288920000000 1279288920000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [29626]) Branching to 'domish-xmlns-spaces-4503'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun removed owner|
|-|-|

As far as I can tell, spaces are allowed here and _change the value_.  See http://www.w3.org/TR/REC-xml-names/#NSNameComparison which specifies the URI to be the normalized value of the attribute, and http://www.w3.org/TR/REC-xml/#AVNormalize which specifies how to normalize values.  In particular:

    For a white space character (#x20, #xD, #xA, #x9), append a space character (#x20) to the normalized value.

So

    "!http://www.xmpp.org/extensions/xep-0084.html#ns-data"

and

    "!http://www.xmpp.org/extensions/xep-0084.html#ns-data "

represent different namespaces.  It seems quite unlikely to me that someone intentionally specified the latter value rather than the former, but it seems that the specification forces us to go along with their mistake.

In any case, it seems that the solution of just being more careful about how the values are split up is the correct solution.  It will preserve spaces in the value and avoid the later `TypeError`.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/domish-xmlns-spaces-4503)

**Comment by twisted-trac** [comment]
|<img alt="Screwtape's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|Screwtape set owner to @exarkun|
|-|-|

For some reason (possibly age), the build results mention that some tests failed but don't provide logs that show *which* tests failed, so I'm assuming they're known-intermittent failures (otherwise you probably wouldn't have put this patch up for review).

Also, the reference on "how to normalize values" has extra whitespace munging steps for attributes whose type "is not CDATA", but I can't find any reference to what the type of namespace values is. "All attributes for which no declaration has been read SHOULD be treated by a non-validating processor as if declared CDATA", so I guess CDATA is a reasonable assumption.

Let's merge it!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [29852]) Merge domish-xmlns-spaces-4503

Author: michich, exarkun
Reviewer: screwtape
Fixes: [#4503](https://github.com/twisted/twisted/issues/4503)

Be careful when splitting namespace URIs from element and attribute
names in the XMPP DOM creation code so that namespace URIs containing
 spaces don't trigger an unhandled exception.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
